[juju] Get Ready for 2.0 and make it less likely to get too much

BryanQuigley opened the pull request:
The package name juju is no longer the only way to install juju,
so replace with checking /usr/bin/juju|juju-run.

Juju-run is on machines deployed with juju.  Logging still occurs
there in same pattern - /var/log/juju/machinename.log.

Juju "Local" logs were never seen in the wild and has been removed.

Check that /var/log/juju/ exists before trying to access it.
Check that /usr/bin/juju exists before trying to run it.
Check that Juju's mongodb exists before trying to parse it.

/var/lib/juju/ contains a database that can grow to many GB, let's
not include that by default.

Also explicitly specified juju formats tabular and json (in 2.0
tabular is the default and in 1.25 json is).
And we also now explicitly capture the version.

Fix up contrainsts and get-config/get so it works for both versions.

Signed-off-by: Bryan Quigley <bryan.quigley@canonical.com>

bmr-cymru commented on sos/plugins/juju.py:
```diff
@@ -76,37 +86,40 @@ def export_mongodb(self):
+                for filename in os.listdir("/var/log/juju/"):
+                    if filename.endswith(".log"):
+                        fullname = "/var/log/juju/" + filename
+                        self.add_copy_spec_limit(fullname, sizelimit=limit)
```
What's the point of doing this in a loop vs. just `add_copy_spec_limit()` with a glob for files matching "*.log"?

BryanQuigley replied:
If I use just add_copy_spec_limit("/var/log/juju/*.log",sizelimit=limit) it will only collect to limit of all the logs in that directory.  (Which in Juju's case will always only collect the all-machines.log because it is always biggest).

If I do each one separately they have their own limit.

bmr-cymru replied:
OK - I think we can do something better that makes it more clear that's what's going on. Let me think about it.

bmr-cymru commented on sos/plugins/juju.py:
```diff
@@ -76,37 +86,40 @@ def export_mongodb(self):
+            if(os.path.isdir("/var/log/juju/")):
+                for filename in os.listdir("/var/log/juju/"):
+                    if filename.endswith(".log"):
+                        fullname = "/var/log/juju/" + filename
```
Nack. You need to use `os.path.join` if you are composing paths from strings.

bmr-cymru commented on sos/plugins/juju.py:
```diff
@@ -76,37 +86,40 @@ def export_mongodb(self):
                 "/var/log/juju",
-                "/var/log/juju-*"
+                "/var/log/juju-*",
+                "/var/lib/juju"
```
What's in this directory? It could use a comment as locations in `/var` sometimes wind up with considerable amounts of data. If that's never going to be the case here as comment would be good to explain that.

bmr-cymru replied:
OK so this was a move within the file from the previous version - still could use a comment though as it's in `/var`.

bmr-cymru commented on sos/plugins/juju.py:
```diff
@@ -76,37 +86,40 @@ def export_mongodb(self):
+            for service in self.get_deployed_services():
+                self.add_cmd_output("juju get-config {}".format(service))
+            for service in self.get_deployed_services():
+                self.add_cmd_output("juju get-constraints {}".format(service))
```
Nack. This is running the exact same `juju status` command three times in a row and then parsing the result!

Just cache the returned value and use it in the subsequent loops:

```python
    services = self.get_deployed_services()
    for service in services:
        # do stuff
    for service in services:
       # do other stuff
```

Or better yet, since these loops are all iterating over the same thing, just get rid of them and have one:

```python
    for service in self.get_deployed_services():
        self.add_cmd_output("juju get {}".format(service))
        self.add_cmd_output("juju get-config {}".format(service))
        self.add_cmd_output("juju get-constraints {}".format(service))
```

battlemidget replied:
You really only need `juju -v status --format=json` then you can perform the rest of your commands on that cached value.

bmr-cymru replied:
You could even make it a single a_c_o call now:

```python
    for service in self.get_deployed_services():
        self.add_cmd_output([
            "juju get {}".format(service),
            "juju get-config {}".format(service),
            "juju get-constraints {}".format(service)
        ])
```

bmr-cymru replied:
@battlemidget that's what `get_deployed_services()` does (`juju status --format json` followed a json.load).

bmr-cymru commented on sos/plugins/juju.py:
```diff
@@ -76,37 +86,40 @@ def export_mongodb(self):
-
-        for service in self.get_deployed_services():
-            self.add_cmd_output("juju get {}".format(service))
+        if(os.path.exists("/usr/bin/juju")):
```
This doesn't need to be in an if block (and generally we do not check for command presence in sos unless it is absolutely necessary). Just attempt the `add_cmd_output()` and if the command doesn't exist it won't be collected. Similarly with the `get_deployed_services()` stuff - just have the function return an empty list if it can't run the command and the loop bodies disappear.

bmr-cymru commented on sos/plugins/juju.py:
```diff
@@ -43,7 +43,17 @@ class Juju(Plugin, UbuntuPlugin):
+            juju = os.path.exists("/usr/bin/juju")
+            jujurun = os.path.exists("/usr/bin/juju-run")
+            if juju or jujurun:
+                return True
```
Nack. Just define a paths list including these path locations.

BryanQuigley commented:
Added note about why not to use add_copy_spec_limit for /var/log/juju. Open to other options of course.
Now using os.path.join
Removed one juju status run by caching it.  I do want to explicitly capture both tabular and json.  
Hugely simplified collecting of get/get-config/ per conversation. 
Uses proper files=  checker.

bmr-cymru commented on sos/plugins/juju.py:
```diff
@@ -76,37 +77,42 @@ def export_mongodb(self):
-            self.add_copy_spec_limit("/var/log/juju-*/all-machines.log",
-                                     sizelimit=limit)
+            # We need this because we want to collect to the limit of all
+            # *.logs in the directory.
```
:+1:

bmr-cymru commented on sos/plugins/juju.py:
```diff
@@ -76,37 +77,42 @@ def export_mongodb(self):
+                "juju get {}".format(service),
+                "juju get-config {}".format(service),
+                "juju get-constraints {}".format(service)
+            ])
```
:+1:

bmr-cymru commented on sos/plugins/juju.py:
```diff
@@ -76,37 +77,42 @@ def export_mongodb(self):
+                "/var/lib/juju"
+                # /var/lib/juju used to be in the default capture moving here
+                # because it usually was way to big.  However, in most cases
+                # you want all logs you want this too.
```
:+1:

bmr-cymru commented:
Thanks Bryan - I've had a quick scan and everything looks fine so far. I've got some time set aside tomorrow for sos pulls so I'll give this a proper look then with a view to merging.

bmr-cymru commented:
@BryanQuigley generally: yes :-) I forgot to take it off this one.

bmr-cymru commented:
Closed in commit 42e3eac.
